### PR TITLE
SDK-100: adding support for Template: create, edit, view, list and run

### PR DESCRIPTION
**Pull request by rahul26goyal**

### CODE REVIEW

**Comment by msumit** [comment]
LGTM

File: qds_sdk/template.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+"""
+import json
+import os
+
+from argparse import ArgumentParser
+from qds_sdk.qubole import Qubole
+from qds_sdk.resource import Resource
+
+class TemplateCmdLine:
+
```
Comment (harshshah87) [inline qds_sdk/template.py]: Can you add comment for the class.  Helps in document generation using sphinx.  More important would be the class Template. 
Also comment on each important function(about the input parameters and output format if any) within Template class which you feel will be used by the sdk-user and needs to be exposed via the python code document.

File: tests/test_template.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        file_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'input_run_template.json')
+        sys.argv = ['qds.py', 'template', 'run', '--id', '14', '--j', file_path]
+        print_command()
+        Connection._api_call = Mock()
+        Connection._api_call.side_effect = template_side_effect
+        qds.main()
+        data = {'input_vars': [{'table': "'accounts'"}]}
+        Connection._api_call.assert_called_with("POST", "command_templates/14/run", data)
+    
+    def test_run_template_with_inine_json(self):
```
Comment (harshshah87) [inline tests/test_template.py]: inline* typo

File: tests/test_template.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+from test_base import print_command
+from test_base import QdsCliTestCase
+
+class TestTemplateCheck(QdsCliTestCase):
+    def test_create_template(self):
+        file_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'input_create_template.json')
+        sys.argv = ['qds.py', 'template', 'create', '--data', file_path]
+        print_command()
+        Connection._api_call = Mock()
+        Connection._api_call.side_effect = template_side_effect
```
Comment (harshshah87) [inline tests/test_template.py]: Why do we have this side_effect? Actually I am completely unaware of what is side_effect in api_call, but why do we need to specify an empty effect?

File: tests/test_template.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        qds.main()
+        data = {"command_type": "HiveCommand", "input_vars": [{"default_value": "\'age\'", "name": "col"},
```
Comment (harshshah87) [inline tests/test_template.py]: Can we read the file instead of hard coding values?

File: tests/input_create_template.json [hunk 5]
```diff
@@ -0,0 +1,26 @@
+{
+    "name" : "tempplate_name",
```
Comment (harshshah87) [inline tests/input_create_template.json]: Is "template_name" a keyword which cannot be used?

File: tests/test_template.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    def test_run_template_with_inine_json(self):
+        sys.argv = ['qds.py', 'template', 'run', '--id', '14', '--j', '[{"table" : "accounts"}]']
+        print_command()
+        Connection._api_call = Mock()
+        Connection._api_call.side_effect = template_side_effect
+        qds.main()
+        data = {'input_vars': [{'table': "'accounts'"}]}
+        Connection._api_call.assert_called_with("POST", "command_templates/14/run", data)
+        
+
```
Comment (harshshah87) [inline tests/test_template.py]: Can you negative test cases. Like no data parameter or id is sent  in some cases.

File: tests/test_template.py (hunk 4 above)
Comment (rahul26goyal) [inline tests/test_template.py]: @harshshah87 Yes..I will validate the data from the input file.

File: tests/input_create_template.json (hunk 5 above)
Comment (rahul26goyal) [inline tests/input_create_template.json]: *typo. "template_name" can be used..

File: tests/test_template.py (hunk 6 above)
Comment (rahul26goyal) [inline tests/test_template.py]: will add negative test cases for this.

File: qds_sdk/template.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+        view = subparsers.add_parser("view", help="To View an existing Template")
+        view.add_argument("--id",dest="id",required=True,help="Id for the Template")
+        view.set_defaults(func=TemplateCmdLine.view)
+        
+        #list 
+        list = subparsers.add_parser("list", help="To List existing Templates")
+        list.add_argument("--per-page", dest="per_page", help="Number of items per page")
+        list.add_argument("--page", dest="page", help="Page Number")
+        list.set_defaults(func=TemplateCmdLine.list)
+
```
Comment (itsvikramagr) [inline qds_sdk/template.py]: add support for clone

File: qds_sdk/template.py [hunk 8]
```diff
@@ -98,31 +123,109 @@ def formatData(inputs):
...
+        Returns:
+            Dictionary containing Command Object details.  
+        """
         conn = Qubole.agent()
         return conn.post(Template.element_path(id + "/run"), data)
     
     @staticmethod
+    def runTemplate(id, data):
+        """
+        Run an existing Template.
```
Comment (harshshah87) [inline qds_sdk/template.py]: Add info that it waits on the results.
Also add the info on return value, that it waits for the command to complete and prints output to stdout.

File: tests/test_template.py [hunk 9]
```diff
@@ -24,67 +24,90 @@ def test_create_template(self):
...
         print_command()
         Connection._api_call = Mock()
-        Connection._api_call.side_effect = template_side_effect
         qds.main()
         data = {'input_vars': [{'table': "'accounts'"}]}
         Connection._api_call.assert_called_with("POST", "command_templates/14/run", data)
         
-        
-        
+    def test_run_template_without_id_data(self):
```
Comment (harshshah87) [inline tests/test_template.py]: Can you add good case of test_run_template.

File: qds_sdk/template.py [hunk 10]
```diff
@@ -98,31 +123,109 @@ def formatData(inputs):
...
+    
+    @staticmethod
     def listTemplates(args):
+        """
+        Fetch existing Templates details.
+        
+        Args:
+            `args`: dictionary containing the value of page number and per_page value
+        Returns:
+            List containging templates details
```
Comment (harshshah87) [inline qds_sdk/template.py]: Typo: Containing template details in Json string?

File: qds_sdk/template.py [hunk 11]
```diff
@@ -2,70 +2,104 @@
...
         with open(args.data) as f:
             spec = json.load(f)
         return Template.editTemplate(args.id, spec)
     
     @staticmethod
-    def execute(args, otherArgs):
+    def submit(args):
+        spec = getSpec(args)
+        res = Template.submitTemplate(args.id, spec)
+        log.info("Submitted Template  with Id: %s, Command Id: %s, CommandType: %s" % (args.id, "12","221"))
```
Comment (harshshah87) [inline qds_sdk/template.py]: Command id "12" is hardcoded?  Shouldn't res have the command id?

File: qds_sdk/template.py (hunk 8 above)
Comment (rahul26goyal) [inline qds_sdk/template.py]: Done

File: tests/test_template.py (hunk 9 above)
Comment (rahul26goyal) [inline tests/test_template.py]: added a test case for run_template as well

File: qds_sdk/template.py (hunk 10 above)
Comment (rahul26goyal) [inline qds_sdk/template.py]: corrected.

File: qds_sdk/template.py (hunk 11 above)
Comment (rahul26goyal) [inline qds_sdk/template.py]: my bad..should print from response.

File: qds_sdk/template.py (hunk 7 above)
Comment (rahul26goyal) [inline qds_sdk/template.py]: Done

**Comment by harshshah87** [comment]
Closing the request as merged into unreleased.

**Pull request closed without merging**